Build fails for an AI service with a polymorphic (`@JsonSubTypes`) return type since Quarkus 3.36.0

## Summary

An `@RegisterAiService` method whose return type is a Jackson polymorphic type
(`@JsonTypeInfo` + `@JsonSubTypes`) makes the
build fail during bytecode recording of `AiServicesRecorder.setMetadata(Map)`.

The thrown exception is a `ClassCastException`
(`TypeVariableImpl cannot be cast to Class`) that has **nothing to do with the real
cause**. It is a secondary failure: the Quarkus recorder hits a serialization error
on the AI-service metadata, tries to build its diagnostic message by calling
`toString()` on the metadata object, and that `toString()` itself throws the CCE.
The original "couldn't serialize property X" message is lost.

So there are two separate defects:

- **A — the mask.** `AiServiceClassCreateInfo` is a record that always carries an
  `InputGuardrailsLiteral` and an `OutputGuardrailsLiteral` component — populated with
  **empty** literals even when the method declares no guardrails. Both literals inherit
  a `toString()` from `AnnotationLiteral` that throws `ClassCastException`
  unconditionally, so `AiServiceClassCreateInfo.toString()` always throws — guardrails
  or not. This destroys every recorder diagnostic for AI-service metadata and makes the
  real failure undiagnosable.
- **B — the trigger.** A polymorphic `@JsonSubTypes` return type produces
  `AiServiceClassCreateInfo` / `AiServiceMethodCreateInfo` metadata that the Quarkus
  bytecode recorder cannot serialize at build time.

**How the two relate.** A is always present but dormant — `toString()` on the metadata
is never called on the happy path. The recorder only calls it from a `catch` block, to
build the "couldn't load object … on object '<param>'" diagnostic after a serialization
failure. So you need a serialization failure (B) to enter that error path, and once
there A corrupts the message. With a non-polymorphic return type, B never fires, the
catch block is never reached, and A stays invisible — which is why the crash looks
specific to the polymorphic return type even though A is unconditional. Any other
metadata-serialization failure would surface the same mask.

Fixing **A** is small and high value on its own: it lets the real **B** error
surface so it can be diagnosed.

## Environment

- **First broken version: Quarkus platform BOM `3.36.0`.** Reproduces on every
  `3.36.x` (`3.36.0`, `3.36.1`, `3.36.2`). Quarkus `3.35.x` builds clean.
- `3.36.0` is the first platform release that ships `quarkus-langchain4j` `1.10.0`
  (the `3.35.x` line shipped `1.9.x`), so the regression is the 1.9.x → 1.10.0 jump.
- `quarkus-langchain4j-core` / `quarkus-langchain4j-bedrock` `1.10.0`.
- `dev.langchain4j:langchain4j-core` `1.14.1`.
- JDK 21 and JDK 25 (reproduces on both).

langchain4j 1.14.0 added "Support polymorphic return types and tool parameters"
(langchain4j#5060); quarkus-langchain4j 1.10.0 records the resulting structured-output
schema metadata at build time, which is where B surfaces.

## Stack trace (as seen)

```
Build step io.quarkus.deployment.steps.MainClassBuildStep#build threw an exception:
java.lang.RuntimeException: Failed to record call to method public void
io.quarkiverse.langchain4j.runtime.AiServicesRecorder.setMetadata(java.util.Map)
    at io.quarkus.deployment.recording.BytecodeRecorderImpl.writeBytecode(BytecodeRecorderImpl.java:507)
    ...
Caused by: java.lang.RuntimeException: java.lang.ClassCastException:
class sun.reflect.generics.reflectiveObjects.TypeVariableImpl cannot be cast to class java.lang.Class
    at io.quarkus.deployment.recording.BytecodeRecorderImpl.loadComplexObject(BytecodeRecorderImpl.java:1479)
    ...
Caused by: java.lang.ClassCastException:
class sun.reflect.generics.reflectiveObjects.TypeVariableImpl cannot be cast to class java.lang.Class
    at jakarta.enterprise.util.AnnotationLiteral.getTypeParameter(AnnotationLiteral.java:113)
    at jakarta.enterprise.util.AnnotationLiteral.annotationType(AnnotationLiteral.java:125)
    at jakarta.enterprise.util.AnnotationLiteral.toString(AnnotationLiteral.java:136)
    at io.quarkiverse.langchain4j.runtime.aiservice.AiServiceClassCreateInfo.toString(AiServiceClassCreateInfo.java:11)
    at java.base/java.lang.String.valueOf(String.java:4465)
    at io.quarkus.deployment.recording.BytecodeRecorderImpl.loadComplexObject(BytecodeRecorderImpl.java:1436)
    ...
```

 That message construction calls
`AiServiceClassCreateInfo.toString()`, which throws the CCE. The actual
property-load failure (the real cause B) never gets reported.

## Reproducer

https://github.com/aliaksei-dudko/quarkus-bug

The minimal moving parts:

```java
// Polymorphic return type — the trigger.
@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
@JsonSubTypes({
    @JsonSubTypes.Type(TextBlock.class),
    @JsonSubTypes.Type(TableBlock.class),
})
public interface ResponseBlock {}

public record ChatResponse(List<ResponseBlock> blocks) {}

@RegisterAiService(chatLanguageModelSupplier = DummyChatModelProducer.class)
@ApplicationScoped
public interface EchoAgent {
    ChatResponse chat(@UserMessage String message);   // no guardrails, no tools, no MCP
}
```

`DummyChatModelProducer` returns an in-memory `ChatModel`

Isolation of the trigger:

- The crash needs **only** the polymorphic return type.
- Change the return type to a non-polymorphic type (e.g. `String`, or a record with
  no `@JsonSubTypes`) and the build passes.

### Root cause — A (the mask)

Neither `InputGuardrailsLiteral` nor `OutputGuardrailsLiteral` extends
`AnnotationLiteral` directly. They share an intermediate generic class:

```java
public abstract sealed class ClassProvidingAnnotationLiteral<A extends Annotation, G extends Guardrail>
        extends AnnotationLiteral<A> permits InputGuardrailsLiteral, OutputGuardrailsLiteral { ... }

public final class OutputGuardrailsLiteral
        extends ClassProvidingAnnotationLiteral<OutputGuardrails, OutputGuardrail>
        implements OutputGuardrails { ... }
```

`AnnotationLiteral.annotationType()` resolves the annotation type by:

1. `getAnnotationLiteralSubclass(getClass())` — walks up to the class whose direct
   superclass is `AnnotationLiteral`. For both literals that is
   `ClassProvidingAnnotationLiteral`, not the concrete literal class.
2. `getTypeParameter(ClassProvidingAnnotationLiteral.class)` — reads
   `getGenericSuperclass()`, which is `AnnotationLiteral<A>`, and casts its single
   actual type argument to `Class`.

That type argument is the **type variable `A`**, a `TypeVariableImpl`, not a concrete
`Class`. The cast throws `ClassCastException`. Because `toString()`, `equals()`, and
`hashCode()` on `AnnotationLiteral` all call `annotationType()`, every one of them
throws for both literals — **unconditionally**, regardless of whether any guardrail is
configured (the no-arg constructor still builds an empty literal).

`AiServiceClassCreateInfo` is a record:

```java
public record AiServiceClassCreateInfo(Map<String, AiServiceMethodCreateInfo> methodMap, String implClassName,
        InputGuardrailsLiteral inputGuardrails, OutputGuardrailsLiteral outputGuardrails) { }
```

Its auto-generated `toString()` renders `inputGuardrails` and `outputGuardrails` in
order, so `AiServiceClassCreateInfo.toString()` always throws the CCE. In the
guardrail-free reproducer it is `inputGuardrails` — the first literal component — that
throws first; with guardrails configured the same break happens. The Quarkus recorder
calls this `toString()` only on the error path (building a "couldn't serialize"
diagnostic), so the CCE replaces the message that would have named the real culprit.

**Fix for A:** make the literals resolve their annotation type without relying on
`AnnotationLiteral.getTypeParameter` walking through a type variable. Options:

- Override `annotationType()` in `InputGuardrailsLiteral` / `OutputGuardrailsLiteral`
  to return `InputGuardrails.class` / `OutputGuardrails.class` directly.
- Or drop the intermediate `ClassProvidingAnnotationLiteral` so each literal extends
  `AnnotationLiteral<ConcreteAnnotation>` directly with a concrete type argument.

Either makes `toString()` safe and unblocks the recorder's real diagnostic.

## Root cause — B (the trigger, masked by A)

With A still in place we cannot read B's exact failing property — that is the whole
point of fixing A first. From the reproducer's isolation (only the polymorphic return
type is needed; a non-polymorphic return type builds clean), the failing property is
in the return-type-derived metadata on `AiServiceMethodCreateInfo`, most likely
`responseSchemaInfo.structuredOutputSchema` (`Optional<JsonSchema>`): the build-time
structured-output schema generated for the `@JsonSubTypes` hierarchy contains an
object the Quarkus `BytecodeRecorderImpl` cannot serialize (a raw `java.lang.reflect.Type`
or a schema node holding one). Once A is fixed, the recorder message will name the
exact property and type — please re-run the reproducer after the A fix to capture it.

## Why it matters

- Any app on quarkus-langchain4j 1.10.0 that returns a polymorphic structured-output
  type from an AI service fails to build, with an error that points at CDI annotation
  literals and gives no hint about the return type.
- Defect A makes B (and any other AI-service metadata serialization failure)
  effectively undiagnosable, because the recorder's own diagnostic throws.

--- a/core/deployment/src/main/java/io/quarkiverse/langchain4j/deployment/ObjectSubstitutionUtil.java
+++ b/core/deployment/src/main/java/io/quarkiverse/langchain4j/deployment/ObjectSubstitutionUtil.java
@@ -1,5 +1,6 @@
 package io.quarkiverse.langchain4j.deployment;
 
+import dev.langchain4j.model.chat.request.json.JsonAnyOfSchema;
 import dev.langchain4j.model.chat.request.json.JsonArraySchema;
 import dev.langchain4j.model.chat.request.json.JsonBooleanSchema;
 import dev.langchain4j.model.chat.request.json.JsonEnumSchema;
@@ -10,6 +11,7 @@ import dev.langchain4j.model.chat.request.json.JsonRawSchema;
 import dev.langchain4j.model.chat.request.json.JsonReferenceSchema;
 import dev.langchain4j.model.chat.request.json.JsonSchema;
 import dev.langchain4j.model.chat.request.json.JsonStringSchema;
+import io.quarkiverse.langchain4j.runtime.substitution.JsonAnyOfSchemaObjectSubstitution;
 import io.quarkiverse.langchain4j.runtime.substitution.JsonArraySchemaObjectSubstitution;
 import io.quarkiverse.langchain4j.runtime.substitution.JsonBooleanSchemaObjectSubstitution;
 import io.quarkiverse.langchain4j.runtime.substitution.JsonEnumSchemaObjectSubstitution;
@@ -30,6 +32,8 @@ final class ObjectSubstitutionUtil {
     static void registerJsonSchema(RecorderContext recorderContext) {
         recorderContext.registerSubstitution(JsonSchema.class, JsonSchemaObjectSubstitution.Serialized.class,
                 JsonSchemaObjectSubstitution.class);
+        recorderContext.registerSubstitution(JsonAnyOfSchema.class, JsonAnyOfSchemaObjectSubstitution.Serialized.class,
+                JsonAnyOfSchemaObjectSubstitution.class);
         recorderContext.registerSubstitution(JsonArraySchema.class, JsonArraySchemaObjectSubstitution.Serialized.class,
                 JsonArraySchemaObjectSubstitution.class);
         recorderContext.registerSubstitution(JsonBooleanSchema.class, JsonBooleanSchemaObjectSubstitution.Serialized.class,

--- a/core/deployment/src/test/java/io/quarkiverse/langchain4j/test/JsonAnyOfSchemaSubstitutionTest.java
+++ b/core/deployment/src/test/java/io/quarkiverse/langchain4j/test/JsonAnyOfSchemaSubstitutionTest.java
@@ -1,0 +1,78 @@
+package io.quarkiverse.langchain4j.test;
+
+import java.util.function.Supplier;
+
+import jakarta.enterprise.context.control.ActivateRequestContext;
+import jakarta.inject.Inject;
+
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import com.fasterxml.jackson.annotation.JsonSubTypes;
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
+import com.fasterxml.jackson.annotation.JsonTypeName;
+
+import dev.langchain4j.data.message.AiMessage;
+import dev.langchain4j.model.chat.ChatModel;
+import dev.langchain4j.model.chat.request.ChatRequest;
+import dev.langchain4j.model.chat.response.ChatResponse;
+import dev.langchain4j.service.UserMessage;
+import io.quarkiverse.langchain4j.RegisterAiService;
+import io.quarkus.test.QuarkusUnitTest;
+
+public class JsonAnyOfSchemaSubstitutionTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest unitTest = new QuarkusUnitTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class));
+
+    @Inject
+    MyAiService ai;
+
+    @Test
+    @ActivateRequestContext
+    void aiServiceWithJsonSubTypesShouldBuild() {
+        ai.chat("hello");
+    }
+
+    public record MyResponse(java.util.List<Block> blocks) {
+    }
+
+    @JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type")
+    @JsonSubTypes({
+            @JsonSubTypes.Type(TextBlock.class),
+            @JsonSubTypes.Type(NumberBlock.class),
+    })
+    public interface Block {
+    }
+
+    @JsonTypeName("TEXT")
+    public record TextBlock(String content) implements Block {
+    }
+
+    @JsonTypeName("NUMBER")
+    public record NumberBlock(int value) implements Block {
+    }
+
+    @RegisterAiService(chatLanguageModelSupplier = MyChatModelSupplier.class)
+    public interface MyAiService {
+        @UserMessage("Dummy")
+        MyResponse chat(String msg);
+    }
+
+    public static class MyChatModelSupplier implements Supplier<ChatModel> {
+        @Override
+        public ChatModel get() {
+            return new ChatModel() {
+                @Override
+                public ChatResponse doChat(ChatRequest request) {
+                    return ChatResponse.builder()
+                            .aiMessage(new AiMessage("{\"blocks\":[]}"))
+                            .build();
+                }
+            };
+        }
+    }
+}

--- a/core/runtime/src/main/java/io/quarkiverse/langchain4j/runtime/aiservice/AiServiceClassCreateInfo.java
+++ b/core/runtime/src/main/java/io/quarkiverse/langchain4j/runtime/aiservice/AiServiceClassCreateInfo.java
@@ -10,4 +10,9 @@ import io.quarkiverse.langchain4j.guardrails.OutputGuardrailsLiteral;
  */
 public record AiServiceClassCreateInfo(Map<String, AiServiceMethodCreateInfo> methodMap, String implClassName,
         InputGuardrailsLiteral inputGuardrails, OutputGuardrailsLiteral outputGuardrails) {
+
+    @Override
+    public String toString() {
+        return implClassName;
+    }
 }

--- a/core/runtime/src/main/java/io/quarkiverse/langchain4j/runtime/substitution/JsonAnyOfSchemaObjectSubstitution.java
+++ b/core/runtime/src/main/java/io/quarkiverse/langchain4j/runtime/substitution/JsonAnyOfSchemaObjectSubstitution.java
@@ -1,0 +1,30 @@
+package io.quarkiverse.langchain4j.runtime.substitution;
+
+import java.util.List;
+
+import dev.langchain4j.model.chat.request.json.JsonAnyOfSchema;
+import dev.langchain4j.model.chat.request.json.JsonSchemaElement;
+import io.quarkus.runtime.ObjectSubstitution;
+import io.quarkus.runtime.annotations.RecordableConstructor;
+
+public class JsonAnyOfSchemaObjectSubstitution
+        implements ObjectSubstitution<JsonAnyOfSchema, JsonAnyOfSchemaObjectSubstitution.Serialized> {
+    @Override
+    public Serialized serialize(JsonAnyOfSchema obj) {
+        return new Serialized(obj.description(), obj.anyOf());
+    }
+
+    @Override
+    public JsonAnyOfSchema deserialize(Serialized obj) {
+        return JsonAnyOfSchema.builder()
+                .description(obj.description)
+                .anyOf(obj.anyOf)
+                .build();
+    }
+
+    public record Serialized(String description, List<JsonSchemaElement> anyOf) {
+        @RecordableConstructor
+        public Serialized {
+        }
+    }
+}

